main/src/main/scala/sbt/Defaults.scala

```diff
@@ -739,7 +743,17 @@ object Defaults extends BuildCommon {
   def runnerInit: Initialize[Task[ScalaRun]] = Def.task {
     val tmp = taskTemporaryDirectory.value
     val si = scalaInstance.value
-    if (fork.value) new ForkRun(forkOptions.value) else new Run(si, trapExit.value, tmp)
+    val s = streams.value
+    val options = javaOptions.value
+    if (fork.value) {
+      s.log.debug(s"javaOptions: $options")
+      new ForkRun(forkOptions.value)
+    } else {
+      if (options.nonEmpty) {
+        s.log.warn("javaOptions will be ignored, fork is set to false")
```

Comment:
Ok, it looks like the best you can do is depend on the `sbt.Keys.configuration` and grab the currently active configuration.  We also need you to grab the scoped "task", ideally, so you can tell users where the bad setting is.